Add natural sort functions for wxString

PBfordev opened the pull request:
This PR builds on #780, attempting to fix some possible deficiencies in that PR.

I have not updated the documentation yet, as this PR may be rejected for the reasons stated in my last comment in #780.

The code also needs to account for `wxUSE_REGEX == 0` but I do not know how to best do it.

Rocketmagnet committed:
Added Natural Sort functions to wxString.

Rocketmagnet committed:
Fixing build errors. Adding test for wxUSE_REGEX.

Rocketmagnet committed:
Moved natural sort functions from wxString to wxArrayString.

Rocketmagnet committed:
Requested changes.

Rocketmagnet committed:
Requested changes.

Rocketmagnet committed:
Requested changes.

Rocketmagnet committed:
Adding tests for wxNaturalStringSortAscending().

Rocketmagnet committed:
Using StrCmpLogicalW() under MSW.

Rocketmagnet committed:
Trying to solve build errors.

Rocketmagnet committed:
Adding WXDLLIMPEXP_BASE.

Rocketmagnet committed:
Adding WXDLLIMPEXP_BASE to arrstr.h too.

Rocketmagnet committed:
Renaming functions.

Rocketmagnet committed:
using wc_str() instead of c_str().

Rocketmagnet committed:
Fixinf documentation.

Rocketmagnet committed:
Removing case /noCase options.

Rocketmagnet committed:
Putting WXDLLIMPEXP_BASE back in.

Rocketmagnet committed:
Reverting whitespace. Removing deciamal point functionality.

Rocketmagnet committed:
Reverting whitespace.

Rocketmagnet committed:
Handling value as int. Layout fixes. Typo.

Rocketmagnet committed:
Handling value as int, forgot about the conversion.

Rocketmagnet committed:
Correcting comment.

Rocketmagnet committed:
Modifying comment.

Rocketmagnet committed:
Modifying comment, again.

Rocketmagnet committed:
__WXMSW__ -> __WINDOWS__

Rocketmagnet committed:
Opening up both Native and WX versions. Adding GNU version. Testing only WX version.

Rocketmagnet committed:
Opening up both Native and WX versions. Adding GNU version. Testing only WX version.

Rocketmagnet committed:
Removing native GNU natural sort strverscmp().

Rocketmagnet committed:
Testing #define _WIN32_IE 0x0501 to enable StrCmpLogicalW().

Rocketmagnet committed:
Testing using __MINGW32_TOOLCHAIN__ to disable StrCmpLogicalW().

Rocketmagnet committed:
Declaring StrCmpLogicalW() in arrstr.cpp #ifdef __MINGW32_TOOLCHAIN__.

Rocketmagnet committed:
Try removing LWSTDAPI_().

Rocketmagnet committed:
Using extern "C".

Rocketmagnet committed:
Using extern "C" only once!

PBfordev committed:
Merge commit 'refs/pull/780/head' of https://github.com/wxWidgets/wxWidgets into natural-sort

PBfordev committed:
Refactor the natural string compare and sort algorithm

Add a new string fragment type for whitespace and punctuation which needs
to be assessed separately from letters and symbols.

Use wxUint64 instead of long for storing the value for numeric fragment.

Use collate instead of compare for non-numeric fragments.

Change names for the public comparison functions: wxWidgets provided function
is now named wxCmpGenericNatural() and for common public use is wxCmpNatural()
which calls a native function in wxMSW and wxCmpGenericNatural() elsewhere.

Make some other changes to simplify and possibly speed up the code.

PBfordev committed:
Update strings in tests

PBfordev committed:
Add headers needed for wxMin() and wxStrcoll_String()

In src/arrstr.cpp add wx/utils.h and wx/wxcrt.h includes.

vadz commented on include/wx/arrstr.h:
```diff
@@ -42,12 +42,30 @@ wxDictionaryStringSortAscending(const wxString& s1, const wxString& s2)
+int wxCMPFUNC_CONV wxCmpNatural(const wxString& s1, const wxString& s2);
+
+WXDLLIMPEXP_BASE
+int wxCMPFUNC_CONV wxCmpGenericNatural(const wxString& s1, const wxString& s2);
```
Nitpicking, but I wonder if this shouldn't be rather called `wxCmpNaturalGeneric()` as we seem to be putting `Generic` either in the beginning or at the end, but rarely in the middle AFAICS.

PBfordev replied:
Makes sense, my intention was probably to maintain wxCmp prefix to indicate to which function family it belongs and the generic classes AFAIK start with "Generic", i.e., `wxGenericDirCtrl`.

Will change to your suggestion.

vadz commented on interface/wx/arrstr.h:
```diff
@@ -403,11 +407,98 @@ int wxDictionaryStringSortAscending(const wxString& s1, const wxString& s2);
+             wxStringSortAscending(),
+             wxDictionaryStringSortAscending()
+        
+        @since 3.1.2  
```
```suggestion
        @since 3.1.4
```
(also elsewhere)

PBfordev replied:
As I wrote in the PR, I have not touched the docs yet, as I thought this PR had little chances to get accepted.

vadz commented on interface/wx/arrstr.h:
```diff
@@ -403,11 +407,98 @@ int wxDictionaryStringSortAscending(const wxString& s1, const wxString& s2);
+             
+        @since 3.1.2
+    */
+int wxCMPFUNC_CONV wxCmpNaturalNative(const wxString& s1, const wxString& s2);
```
This function doesn't seem to exist, so shouldn't be documented...

PBfordev replied:
See above.

vadz commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +730,212 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+
+namespace    // helpers needed by wxCmpGenericNatural()
+{
+    // Used for comparison of string parts
```
Trivial, but we don't indent `namespace` contents to avoid indenting big blocks of text.

PBfordev replied:
Will fix, I did not change it from the original PR.

vadz commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +730,212 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+                fragment.type = wxStringFragment::SpaceOrPunct;
+                fragment.text = text.Left(length);
+            }
+        } else
```
```suggestion
        }
        else
```
Just Another Style Nitpick

vadz commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +730,212 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+            }
+        } else
+        {
+            wxFAIL_MSG("text doesn't start with any recognizable fragment type");
```
I don't think we should give assert failures for any kind of user-defined input. Not sure what is supposed to happen here, but it looks like we could just remove the `else` branch entirely.

PBfordev replied:
I wrote it like this because this is not a failure of input (unlike with e.g. IO op). If we get there it is a wxWidgets bug as we should always get recognize the string as one of the given fragments.

However, you are probably correct that is it not a user (application) programmer fault so he should not get an assert for wxWidgets bug. Will change.

vadz commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +730,212 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+    {
+        int result = 1;
+
+        if ( lhs.type == wxStringFragment::Empty )
```
I'd prefer to write this as `switch ( lhs.type )` instead of a sequence of `if` statements.

This is less trivial than other comments, as this can guarantee that the switch is exhaustive.

PBfordev replied:
I thought of it too originally, but that would be one giant switch, including nested switches but I will try do that.

vadz commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +730,212 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+                case wxStringFragment::LetterOrSymbol:
+                    result = -1;
+                    break;
+                default:
+                    wxFAIL_MSG("Unknown string fragment type");
```
For the same exhaustiveness reasons I'd remove this because compiler wouldn't warn about missing case if we ever add another fragment type with the `default` present.

(also below)

PBfordev replied:
Will do.

vadz reviewed (commented):
Thanks, this looks broadly good to me and the remaining small changes should be very simple to make.

Concerning `wxUSE_REGEX`, I think it would be fine to just fall back on `wxStrcmp()` in this case, we just would need to mention this in the documentation (and also disable unit test, i.e. add a warning about skipping it, in this case).

Thanks again for reviving this, it would be nice to finally see it merged!

vadz commented:
I may be missing something here, but why are you so pessimistic about the chance of this being merged? I don't see any really bad problems with these changes. Of course, they're not perfect, but then nothing is and things could be improved later, while for now this already adds a wrapper for a useful (in the context of UI development) MSW function with a reasonable fallback under the other platforms, so I think it should be fine to merge it... Unless, once again, I simply oversaw some critical problem?

PBfordev commented:
>why are you so pessimistic about the chance of this being merged? 

My innate pessimism aside, the issues I see (none is critical):

1. I cannot guarantee there is not a bug there, the algorithm seems simple enough, but even the original PR had it wrong.
2. Sorting like this is very inefficient since we do not provide a sort function but only a comparison function. Hence string transformations must be done during each comparison, i.e., possibly many times for each string. OTOH, the function should probably not be used for large amounts of strings.
3. The "fragmentization" algorithm does not properly handle control characters. OTOH, this function should be used for strings that should not have such characters; moverover, ASCII tabs and new lines which are probably most common control characters for strings presented to the user are included in `:space:`. I thought of adding a string fragment type using regex `:cntrl:` but that would probably involve also Unicode control characters which are a can of worms which I did not dare open and left it to the collation function.
4. The original PR was in a limbo for two years.

Anyway, I will address the issues pointed in the review, hopefully by Tuesday at the latest.

PBfordev committed:
Address review comments

PBfordev committed:
Update documentation

PBfordev committed:
Remove incorrect comparisons from tests

"$" is not considered a letter/symbol, but a punctuation.

PBfordev committed:
Fix issues from previous commits

In wxCmpNaturalGeneric() do not use wxStrcmp() which requires
including wxcrt.h; use wxString::compare() instead.

In the wxCmpNaturalGeneric() documentation, fix the typo in wxRegEx
(was wxRegex) and do not refer to undocumented wxStrcmp().

PBfordev commented:
I have addressed the review comments, please let me know if there something more to be done here.

PBfordev committed:
Add missing break in switch case

vadz commented:
Looks good to me, I'm ready to merge this once the CI builds finish. Thanks again!

PBfordev commented:
You probably have a plan for this but I wonder how to merge it? I think, if possible, it would be best to somehow squash it into just two commits: the first containing all the commits by Rocketmagnet, the second with all mine.

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+            {
+                case wxStringFragment::Empty:
+                    result = 0;
+                    break;
```
Why not just `return 0;`? There's nothing being done at the end of the outer switch, and one would need quite a bit of scrolling to find that out.

PBfordev replied:
This is basically so that the code for all cases is analogical. I believe that if we return from here, we should return from all cases.

catalinr replied:
Yes, the same for all `case`s.
I think it will help with readability, rather than anything else.

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+int wxCMPFUNC_CONV wxCmpNaturalGeneric(const wxString& s1, const wxString& s2)
+{
+    wxString lhs(s1);
+    wxString rhs(s2);
```
Since these are copied immediately inside the function, then the arguments can be values themselves, can't they?

PBfordev replied:
I am not sure I do understand. Did you mean to remove the const reference from the function arguments? I would rather keep the function signature same for all `wxCmpXXX() `functions.

catalinr replied:
Ok, I didn't know about other signatures. Consistency should come first, indeed.

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+// use the generic version.
+inline int wxCMPFUNC_CONV wxCmpNatural(const wxString& s1, const wxString& s2)
+{
+    #if defined( __WINDOWS__ )
```
these should probably not be indented

PBfordev replied:
OK, will change.

PBfordev commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+wxStringFragment GetFragment(wxString& text)
+{
+    static const wxRegEx reSpaceOrPunct(wxS("^([[:space:]]|[[:punct:]])+"));
+    static const wxRegEx reDigit(wxS("^[[:digit:]]+"));
```
Makes sense, will change.

PBfordev commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+    // string fragment type.
+    if ( reLetterOrSymbol.Matches(text) )
+    {
+        if ( reLetterOrSymbol.GetMatch(NULL, &length) )
```
In the previous version, there were `wxCHECK()`s but they were all removed, let's see what @vadz has to say here.

vadz replied:
Sorry, I don't remember all the details about the previous versions, but the principle is always the same: it should never be possible to trigger an assert/check failure due to anything happening outside of the program, i.e. user input, IO error etc. They should only happen in case of preconditions or, more rarely, postconditions failure.

So if we can guarantee that the regex here will always match no matter what, then using `wxCHECK` might make sense. OTOH in this particular case we know that `GetMatch()` will trigger an assertion failure anyhow if it fails (because it can only fail if its preconditions are not satisfied, there is no valid failure mode for it), so it doesn't seem really useful to have it.

PBfordev replied:
Actually `GetMatch()` is called only if `Matches()` returns true. So I have no idea what would actually make `GetMatch()` fail here. I have added the if check for `GetMatch()`, it was not in the original PR, perhaps it is useless but I am just to always checking the return value... 

The original `wxCHECK()` here was when `Matches()` for the last possible string fragment type returned false, which should never happen as the regex expression is the negative of the previous two.

PBfordev commented on src/common/arrstr.cpp:
```diff
@@ -721,3 +732,217 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+    wxStringFragment fragment;
+    size_t           length = 0;
+
+    if ( text.empty() )
```
I thought about this when I "took over" the PR but considered their cost negligible, very unC++, I know.

PBfordev committed:
Address comments from the second review

PBfordev committed:
Try harder in wxCmpNaturalGeneric() if wxRegEx is unavailable

If wxRegEx is not available, do not just make a simple string comparison
but perform a case-insensitive collation.

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -27,6 +27,17 @@
 
+#include "wx/wxcrt.h"
+
+#if wxUSE_REGEX
```
These guards are not needed, they are already inside `wx/regex.h`

PBfordev replied:
Sigh. I even looked at other sources how they do it but looked only at src/msw/textctrl.cpp which has
```cpp
#if wxUSE_CLIPBOARD
    #include "wx/clipbrd.h"
    #include "wx/dataobj.h"
#endif
```

PBfordev committed:
Do not guard regex.h include

PBfordev commented:
I have addressed the comments from the last review.

vadz commented:
Thanks again, I've squash-rebased as suggested and pushed it now.
